### PR TITLE
Use storedFieldsSpec to load stored fields for highlighting

### PR DESCRIPTION
**Pull request by romseygeek**
Since #91269, fetch phase subprocessors can report any stored fields that they
need back to the top-level FetchPhase so that all stored fields can be loaded
up front.  This commit switches the unified and plain highlighters to use this
functionality so that highlighting does not need to open stored field readers
twice.

### CODE REVIEW

**Comment by cbuescher** [comment]
@romseygeek just found this sitting in my queue of PRs assigned for review and wanted to say sorry this fell through the cracks. Seems stale at the moment, do you think it should get updated or is it better to close at this point?

**Comment by romseygeek** [comment]
Still worth doing, I think.  I'll update.

File: server/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/CustomUnifiedHighlighterTests.java [hunk 1]
```diff
@@ -38,4 +38,61 @@ public void testSimpleTermHighlighting() throws IOException {
...
+        ParsedDocument doc = mapperService.documentMapper().parse(source("""
+            { "field_source" : "this is some text" }
+            """));
+
+        SearchSourceBuilder search = new SearchSourceBuilder().query(QueryBuilders.termQuery("field", "some"))
+            .highlighter(new HighlightBuilder().field("field"));
+
+        Map<String, HighlightField> highlights = highlight(mapperService, doc, search);
+        assertHighlights(highlights, "field", "this is <em>some</em> text");
+    }
```
Comment (cbuescher) [inline server/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/CustomUnifiedHighlighterTests.java:96]: These tests already pass without the modifications in this PR. Do we have other ways to verify that the new optimization is actually used, e.g. check frequency of stored field readers access? Some asserting mocks etc?

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightUtils.java [hunk 2]
```diff
@@ -43,10 +38,8 @@ public static List<Object> loadFieldValues(
         FetchSubPhase.HitContext hitContext
     ) throws IOException {
         if (fieldType.isStored()) {
-            CustomFieldsVisitor fieldVisitor = new CustomFieldsVisitor(singleton(fieldType.name()), false);
-            hitContext.reader().document(hitContext.docId(), fieldVisitor);
-            List<Object> textsToHighlight = fieldVisitor.fields().get(fieldType.name());
-            return Objects.requireNonNullElse(textsToHighlight, Collections.emptyList());
+            List<Object> values = hitContext.loadedFields().get(fieldType.name());
```
Comment (cbuescher) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightUtils.java:41]: To check my understanding: is this the part where we actually save having to  open stored field readers more than once? Does the hitContext now always already contain the loaded fields?

**Review by cbuescher — COMMENTED** [review]
I took a look and left two questions. Mostly I'm trying to understand through which part of this change unified and plain highlighters make use of the pre-loading. I think I found the spot but want to make sure I understand this right. Also not sure how difficult it would be to test that the optimization actually gets used.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightUtils.java (hunk 2 above)
Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/HighlightUtils.java:41]: Yes, exactly so

**Review by cbuescher — APPROVED** [review]
Thanks for the additional changes, LGTM. Nice solution for the test with wrapping the directory reader.

**Pull request merged**